### ISSUE TITLE
No retry if socket hangs up

### ISSUE THREAD

**Issue by mfonsen** [takomo-io/takomo#121]
It seems that Takomo doesn't handle socket hang up error properly as it doesn't do a retry. 

Perhaps the concurrency/retry logic has changed when the application was upgraded from Takomo 1.x to Takomo 2.x. 

Takomo version: 2.11.0
Takomo command: takomo stacks deploy ...

Please find relevant logs in debug mode below. All stacks prior to the failed one succeeded. There does seem to be considerable amount of retrying going on due to speed of execution as there are no changes to any of the stacks. It seems somewhat random which stack fails so it's not probably due to any individual stack configuration.
```
2020-10-20 10:10:36 +0300 [debug] - Load existing stack with path: /app/eu-west-1/a.yml/eu-west-1, name: app-eu-west-1-dev
2020-10-20 10:10:37 +0300 [debug] - Load existing template summary for stack with path: /app/eu-west-1/a.yml/eu-west-1, name: app-eu-west-1-dev
2020-10-20 10:10:45 +0300 [debug] - Existing template summary for stack found with path: /app/eu-west-1/a.yml/eu-west-1, name: app-eu-west-1-dev
2020-10-20 10:10:46 +0300 [debug] - Existing stack found with path: /app/eu-west-1/a.yml/eu-west-1, name: app-eu-west-1-dev
  - /app/eu-west-1/a.yml/eu-west-1
2020-10-20 10:10:53 +0300 [info ] - /app/eu-west-1/a.yml/eu-west-1 - Preparing stack deployment
2020-10-20 10:10:53 +0300 [debug] - /app/eu-west-1/a.yml/eu-west-1 - Stack config:
2020-10-20 10:10:53 +0300 [debug] - /app/eu-west-1/a.yml/eu-west-1 - Wait 2 dependencies to complete
2020-10-20 10:13:14 +0300 [debug] - /app/eu-west-1/a.yml/eu-west-1 - Dependencies completed successfully
2020-10-20 10:13:14 +0300 [debug] - /app/eu-west-1/a.yml/eu-west-1 - Execute before deploy hooks
2020-10-20 10:13:14 +0300 [debug] - /app/eu-west-1/a.yml/eu-west-1 - About to execute hooks (operation: update, stage: before, status: null)
2020-10-20 10:13:14 +0300 [debug] - /app/eu-west-1/a.yml/eu-west-1 - Found 1 matching hook(s)
2020-10-20 10:13:14 +0300 [debug] - /app/eu-west-1/a.yml/eu-west-1 - Execute hook (name: stack-policy, type: cmd)
2020-10-20 10:13:15 +0300 [debug] - /app/eu-west-1/a.yml/eu-west-1 - Hook (name: stack-policy, type: cmd) executed with success: true, message: Success
2020-10-20 10:13:15 +0300 [debug] - /app/eu-west-1/a.yml/eu-west-1 - Execute before deploy hooks completed
2020-10-20 10:13:15 +0300 [debug] - /app/eu-west-1/a.yml/eu-west-1 - Prepare CloudFormation template
2020-10-20 10:13:15 +0300 [debug] - /app/eu-west-1/a.yml/eu-west-1 - Path to CloudFormation template: /app/aws/templates/template.yml
2020-10-20 10:13:15 +0300 [debug] - /app/eu-west-1/a.yml/eu-west-1 - No template bucket configured
2020-10-20 10:13:15 +0300 [debug] - /app/eu-west-1/a.yml/eu-west-1 - Validate template
2020-10-20 10:13:19 +0300 [debug] - /app/eu-west-1/a.yml/eu-west-1 - Prepare parameters
2020-10-20 10:13:28 +0300 [debug] - /app/eu-west-1/a.yml/eu-west-1 - Prepare tags
2020-10-20 10:13:28 +0300 [info ] - /app/eu-west-1/a.yml/eu-west-1 - Update stack
2020-10-20 10:13:29 +0300 [debug] - /app/eu-west-1/a.yml/eu-west-1 - Wait for stack deploy to complete
2020-10-20 10:14:02 +0300 [error] - /app/eu-west-1/a.yml/eu-west-1 - Failed to wait for stack launch to complete Error [NetworkingError]: socket hang up
    at connResetException (internal/errors.js:617:14)
    at TLSSocket.socketOnEnd (_http_client.js:493:23)
    at TLSSocket.emit (events.js:327:22)
    at TLSSocket.EventEmitter.emit (domain.js:486:12)
    at endReadableNT (_stream_readable.js:1327:12)
    at processTicksAndRejections (internal/process/task_queues.js:80:21) {
  code: 'NetworkingError',
  region: 'eu-west-1',
  hostname: 'cloudformation.eu-west-1.amazonaws.com',
  retryable: true,
  time: 2020-10-20T07:14:02.718Z,
  retryDelay: -1
}
...
Stack path                                      Stack name                        Status     Reason               Time      Message            
----------------------------------------------  --------------------------------  ---------  -------------------  --------  -------------------
...
/app/eu-west-1/a.yml/eu-west-1          app-eu-west-1-dev          FAILED     ERROR                3m 9.4s   socket hang up     
...
Events for failed stacks
------------------------

/app/eu-west-1/a.yml/eu-west-1

  <no events>

```

**Comment by mfonsen** [comment]
Finally after ~10 retries got the test stack deployed. This would hint that retrying might be a good idea.

**Comment by hmeltaus** [comment]
You are right that Takomo should retry requests failing due to network errors.

Retrying logic is currently based on the error code, and `NetworkingError` was not yet in the list of retryable errors. Errors from AWS do contain a `retryable` boolean property, but we can't solely retry based on it because some errors like validation errors have `retryable` set to true, even though it makes no sense to retry them.

The retry logic is actually same in 1.5.x and 2.x but there must be something else that causes sockets to hang up. I haven't encountered socket hang ups myself but I would guess that they'll start to appear once the configuration grows big enough.

Anyway, I release a quick fix for this issue now, and then check if I can find what actually is causing it.

**Comment by mfonsen** [comment]
Excellent! Thank you. I wasn't able to reproduce the issue any more but the fix does make sense. 

It might have been some temporary network issue or something AWS does to protect their service. I have seen this to happen to the extent that once I had to replace use of AWS Parameter Store with AWS Secrets Manager because I was getting these errors in multiple outside networks and found no other solution at that time.

I found one interesting [article](https://medium.com/@andyhu92/interesting-story-about-how-we-solved-a-socket-hang-up-issue-in-aws-lambda-cd467a6febca) on this type of error. It suggests that AWS SDK itself should retry `NetworkingError` behind the scenes. In their case they had accidentally set retries to one which made the problem visible. Maybe the custom retry logic had the same result?
